File: linked_lists/linked_list_test.rb
gem 'minitest'
require 'minitest/autorun'
require 'minitest/pride'

class LinkedListTest < Minitest::Test
  attr_reader :list

  def setup
    @list = LinkedList.new
  end

  def test_nodes_have_data
    skip
    node = Node.new("pizza")
    assert_equal "pizza", node.data
  end

  def test_nodes_have_next_node
    skip
    n1 = Node.new("pizza")
    n2 = Node.new("cats")
    n1.next_node = n2
    assert_equal "cats", n1.next_node.data
    assert_equal Node, n1.next_node.class
  end

  def test_it_accepts_next_node_on_init
    skip
    n1 = Node.new("pizza", Node.new("cats"))
    assert_equal "cats", n1.next_node.data
    assert_equal Node, n1.next_node.class
  end

  def test_it_starts_with_zero_elements
    skip
    assert_equal 0, list.count
  end

  def test_a_new_list_starts_with_nil_head
    skip
    assert_equal nil, LinkedList.new.head
  end

  def test_it_pushes_a_single_element_onto_a_list
    skip
    list.push("pizza")
    assert_equal "pizza", list.head.data
    assert_equal 1, list.count
  end

  def test_it_pushes_two_elements
    skip
    list.push("pizza")
    assert_equal "pizza", list.head.data
    list.push("stromboli")
    assert_equal "stromboli", list.head.next_node.data
  end

  def test_it_pushes_three_elements_onto_a_list
    skip
    list.push("hello")
    assert_equal "hello", list.head.data
    list.push("world")
    assert_equal "world", list.head.next_node.data
    list.push("today")
    assert_equal "world", list.head.next_node.data
    assert_equal "today", list.head.next_node.next_node.data
    assert_equal 3, list.count
  end

  def test_it_adds_four_elements
    skip
    list.push("hello")
    list.push("world")
    list.push("today")
    list.push("pizza")
    assert_equal "today", list.head.next_node.next_node.data
    assert_equal "pizza", list.head.next_node.next_node.next_node.data
    assert_equal 4, list.count
  end

  def test_it_pops_the_last_element_from_the_list
    skip
    list.push("hello")
    list.push("world")
    list.push("today")
    output = list.pop
    assert_equal "today", output
    assert_equal 2, list.count
  end

  def test_a_popped_element_is_removed
    skip
    list.push("hello")
    output = list.pop
    assert_equal "hello", output
    assert_equal 0, list.count
  end

  def test_it_pops_nil_when_there_are_no_elements
    skip
    assert_nil list.pop
  end

  def test_it_deletes_a_solo_node
    skip
    list.push("hello")
    list.delete("hello")
    assert_equal 0, list.count
  end

  def test_it_does_not_delete_when_the_data_does_not_match
    skip
    list.push("hello")
    list.push("world")
    list.delete("today")
    assert_equal 2, list.count
  end

  def test_it_deletes_a_last_node
    skip
    list.push("hello")
    list.push("world")
    list.push("today")
    list.delete("today")
    assert_equal 2, list.count
  end

  def test_it_deletes_a_middle_node
    skip
    list.push("hello")
    list.push("world")
    list.push("today")
    list.delete("world")
    assert_equal 2, list.count
    assert_equal "today", list.pop
    assert_equal "hello", list.pop
  end

  def test_it_deletes_the_head_when_there_are_more_nodes
    skip
    list.push("hello")
    list.push("world")
    list.push("today")
    list.delete("hello")
    assert_equal 2, list.count
    assert_equal "today", list.pop
    assert_equal "world", list.pop
  end

  def test_it_converts_to_an_array_when_there_are_no_elements
    skip
    assert_equal [], list.to_a
  end

  def test_it_converts_to_an_array_with_several_elements
    skip
    list.push("hello")
    list.push("world")
    list.push("today")
    assert_equal ["hello", "world", "today"], list.to_a
  end

  def test_it_finds_the_last_node
    skip
    list.push("hello")
    list.push("world")
    node = list.last_node
    assert_equal "world", node.data
  end

  def test_a_node_links_to_its_next_element
    skip
    list.push("hello")
    list.push("world")
    assert_equal "world", list.last_node.data
    assert_equal "world", list.head_node.next_node.data
  end

  def test_next_node_for_the_last_node_is_nil
    skip
    list.push("world")
    assert_nil list.last_node.next_node
  end

  def test_find_if_an_element_is_included_in_the_list
    skip
    list.push("hello")
    list.push("world")
    assert_equal true, list.include?("hello")
    assert_equal false, list.include?("bogus")
  end

  def test_find_a_given_node
    skip
    list.push("hello")
    list.push("world")
    list.push("today")

    assert_equal "world", list.find("world").data
    assert_equal "today", list.find("world").next_node.data
  end

  def test_inserts_node_at_arbitrary_position
    skip
    list.push("hello")
    list.push("world")
    list.push("today")

    list.insert(1, "pizza")

    assert_equal 1, list.index("pizza")
    assert_equal ["hello", "pizza", "world", "today"], list.to_a
  end

  def test_inserted_node_is_next_node_for_previous_node
    skip
    list.push("hello")
    list.push("world")
    list.push("today")

    list.insert(1, "pizza")

    assert_equal "world", list.find("pizza").next_node.data
    assert_equal "pizza", list.find("hello").next_node.data
  end

  def test_insert_after_adds_a_node_after_a_given_node
    skip
    list.push("hello")
    list.push("world")
    list.push("today")

    list.insert_after("hello", "pizza")

    assert_equal "world", list.find("pizza").next_node.data
    assert_equal "pizza", list.find("hello").next_node.data
  end

  def test_distance_returns_distance_between_two_nodes
    skip
    list.push("hello")
    list.push("pizza")
    list.push("world")
    list.push("today")
    list.push("tomorrow")
    assert_equal 3, list.distance("hello", "today")
    assert_equal 2, list.distance("pizza", "today")
    assert_equal 2, list.distance("hello", "world")
    assert_equal 4, list.distance("hello", "tomorrow")
    assert_equal 1, list.distance("world", "today")
  end
end

File: linked_lists/javascript/linkedList_test.js
describe('ListNode', function() {
  it('should have data', function() {
    var node = new ListNode('pizza');
    expect(node.data).to.eq('pizza');
  });

  it.skip('should have a default empty nextNode', function() {
    var node = new ListNode('pizza');
    expect(node.nextNode).to.eq(null);
  });

  it.skip('should allow setting a nextNode', function() {
    var n1 = new ListNode('pizza');
    var n2 = new ListNode('cats');
    n1.nextNode = n2;
    expect(n1.nextNode.data).to.eq('cats');
    expect(n1.nextNode instanceof ListNode).to.be.true;
  });

  it.skip('should allow a next node argument on creation', function(){
    var node = new ListNode('pizza', new ListNode('cats'));
    expect(node.nextNode.data).to.eq('cats');
    expect(node.nextNode instanceof ListNode).to.be.true;
  });
});

describe('LinkedList', function() {
  var list;

  beforeEach(function(){
    list = new List();
  });

  it.skip('should start with zero elements', function() {
    expect(list._length).to.eq(0);
  });

  it.skip('should set its default head to null', function(){
    expect(list.head).to.eq(null);
  });

  describe('.push', function(){
    context('with a single element', function(){
      it.skip('should allow push of a single element to a list', function(){
        list.push('pizza');
        expect(list.head.data).to.eq('pizza');
      });

      it.skip('should increment the _length of the list', function(){
        list.push('pizza');
        expect(list._length).to.eq(1);
      });
    });

    context('with multiple elements', function(){
      it.skip('should increment the length count', function(){
        list.push('pizza');
        list.push('stromboli');
        list.push('mushroom');
        expect(list._length).to.eq(3);
      });

      it.skip('should assign the head to the first element pushed', function(){
        expect(list.head).to.eq(null);
        list.push('pizza');
        expect(list.head.data).to.eq('pizza');
        list.push('stromboli');
        expect(list.head.data).to.eq('pizza');
      });

      it.skip('should attach the second element to the first element', function(){
        list.push('pizza');
        list.push('stromboli');
        expect(list.head.nextNode.data).to.eq('stromboli');
      });

      it.skip('should attach nextNodes in sequential order', function(){
        list.push('pizza');
        list.push('stromboli');
        list.push('mushroom');
        list.push('peanutbutter');
        expect(list.head.data).to.eq('pizza');
        expect(list.head.nextNode.data).to.eq('stromboli');
        expect(list.head.nextNode.nextNode.data).to.eq('mushroom');
        expect(list.head.nextNode.nextNode.nextNode.data).to.eq('peanutbutter');
      });
    });
  });

  describe('.pop', function(){
    context('with no elements', function(){
      it.skip('should return null', function(){
        expect(list.pop()).to.eq(null);
      });

      it.skip('should not decrement the _length', function(){
        expect(list._length).to.eq(0);
      });
    });

    context('with one element', function(){
      it.skip('should change the _length', function(){
        list.push('hello');
        var result = list.pop();
        expect(list._length).to.eq(0);
      });

      it.skip('should set the list head to null', function(){
        list.push('hello');
        var result = list.pop();
        expect(list.head).to.eq(null);
      });

      it.skip('should return the last element', function(){
        list.push('hello');
        var result = list.pop();
        expect(result.data).to.eq('hello');
      });
    });

    context('with multiple elements', function(){
      it.skip('should return the last element from the list', function(){
        list.push("hello");
        list.push("new");
        list.push("world");
        list.push("today");

        var output = list.pop();
        expect(output.data).to.eq('today');
      });

      it.skip('should remove the last element from the list', function(){
        list.push("hello");
        list.push("world");
        list.push("today");

        var output = list.pop();
        expect(output.data).to.eq('today');
        expect(list._length).to.eq(2);

        var output2 = list.pop();
        expect(output2.data).to.eq('world');
        expect(output2.nextNode).to.eq(null);
        expect(list._length).to.eq(1);

        var output3 = list.pop();
        expect(output3.data).to.eq('hello');
        expect(output3.nextNode).to.eq(null);
        expect(list._length).to.eq(0);
      });
    });
  });

  describe('.delete', function(){
    context('with one node', function(){
      it.skip('deletes a solo node', function(){
        list.push('hello');
        list.delete('hello');
        expect(list._length).to.eq(0);
        expect(list.head).to.eq(null);
      });

      it.skip('does not perform a delete when a node does not match', function(){
        list.push('hello');
        list.delete('goodbye');
        expect(list._length).to.eq(1);
        expect(list.head.data).to.eq('hello');
      });
    });

    context('with multiple nodes', function(){
      beforeEach(function(){
        list.push('hello');
        list.push('darkness');
        list.push('my');
        list.push('old');
        list.push('friend');
      });

      it.skip('changes the list _.length', function(){
        expect(list.head.nextNode.data).to.eq('darkness');
        expect(list._length).to.eq(5);
        list.delete('friend');
        expect(list._length).to.eq(4);
        list.delete('my');
        expect(list._length).to.eq(3);
        list.delete('happy');
        expect(list._length).to.eq(3);
      });

      it.skip('resets the nextNode property on the node before the deleted node', function(){
        expect(list.head.nextNode.data).to.eq('darkness');
        list.delete('darkness');
        expect(list.head.nextNode.data).to.eq('my');
      });

      it.skip('resets the list.head if deleting the first node', function(){
        expect(list.head.data).to.eq('hello');
        list.delete('hello');
        expect(list.head.data).to.eq('darkness');
      });
    });
  });

  describe('.toArray', function(){
    context('when there are no elements', function(){
      it.skip('converts to an array', function(){
        expect(list.toArray()).to.deep.equal([]);
      });
    });

    context('when there are several elements', function(){
      beforeEach(function(){
        list.push('The');
        list.push('rain');
        list.push('in');
        list.push('Spain');
      });

      it.skip('can convert to an array', function(){
        expect(list.toArray()).to.deep.equal(['The', 'rain', 'in', 'spain']);
      });
    });
  });

  describe('.lastNode', function(){
    context('with several nodes', function(){
      beforeEach(function(){
        list.push('The');
        list.push('rain');
        list.push('in');
        list.push('Spain');
      });

      it.skip('finds the last node', function(){
        expect(list.lastNode().data).to.eq('Spain');
      });
    });

    context('with one node', function(){
      beforeEach(function(){
        list.push('Ahoy!');
      });

      it.skip('finds the only node', function(){
        expect(list.lastNode().data).to.eq('Ahoy!');
      });
    });

    context('with no nodes', function(){
      it.skip('returns null', function(){
        expect(list.lastNode()).to.eq(null);
      });
    });
  });

  describe('.include', function(){
    beforeEach(function(){
      list.push('The');
      list.push('rain');
      list.push('in');
      list.push('Spain');
    });

    it.skip('should return true if node is in list', function(){
      expect(list.include("rain")).to.eq(true);
    });

    it.skip('should return false if node is not in list', function(){
      expect(list.include("nope")).to.eq(false);
    });
  });

  describe('.find', function(){
    beforeEach(function(){
      list.push('oh');
      list.push('hello');
      list.push('world');
    });

    it.skip('should return true the node if node in list', function(){
      var result = list.find("hello");
      expect(result.data).to.eq('hello');
      expect(result.nextNode.data).to.eq('world');
    });

    it.skip('should return null if node is missing', function(){
      var result = list.find("nope");
      expect(result).to.eq(null);
    });
  });

  describe('.index', function(){
    beforeEach(function(){
      list.push('oh');
      list.push('hello');
      list.push('world');
    });

    it.skip('should return expected indexes', function(){
      expect(list.index('oh')).to.eq(0);
      expect(list.index('world')).to.eq(2);
      expect(list.index('nope')).to.eq(null);
    });
  });

  describe('.insert', function(){
    beforeEach(function(){
      list.push('dark');
      list.push('stormy');
    });

    it.skip('should insert nodes', function(){
      expect(list._length).to.eq(2);
      list.insert(1, 'and');
      list.insert(3, 'night');
      expect(list._length).to.eq(4);
      expect(list.toArray).to.deep.equal('dark', 'and', 'stormy', 'night');
    });
  });

  describe('.insertAfter', function(){
    beforeEach(function(){
      list.push('dark');
      list.push('stormy');
    });

    it.skip('should insert nodes after other nodes', function(){
      expect(list._length).to.eq(2);
      list.insertArfter('dark', 'and');
      list.insertAfter('stormy', 'night');
      expect(list._length).to.eq(4);
      expect(list.toArray).to.deep.equal('dark', 'and', 'stormy', 'night');
    });
  });

  describe('.distance', function(){
    beforeEach(function(){
      list.push("hello")
      list.push("pizza")
      list.push("world")
      list.push("today")
      list.push("tomorrow")
    });

    it.skip('should calculate distance between nodes', function(){
      expect(list.distance("hello", "today")).to.eq(3);
      expect(list.distance("pizza", "today")).to.eq(2);
      expect(list.distance("hello", "world")).to.eq(2);
      expect(list.distance("hello", "tomorrow")).to.eq(4);
      expect(list.distance("world", "today")).to.eq(1);
    });
  });
});
